### ISSUE TITLE
How to Minimize Rounding Error in torch.autograd.functional.jacobian?

### ISSUE THREAD

**Issue by QiyaoWei** [pytorch/pytorch#74288]
### 🐛 Describe the bug

Before I start, let me express my sincerest gratitude to issue #49171, in making it possible to take the jacobian wrt all model parameters! A great functionality indeed!

I am raising an issue about the approximation error when the jacobian function goes to high dimensions. In low dimensions, the following code work fine
```
import torch
from torch.autograd.functional import jacobian
from torch.nn.utils import _stateless
from torch import nn
from torch.nn import functional as F
```

```
model = nn.Conv2d(3,1,1)
input = torch.rand(1, 3, 32, 32)
two_input = torch.cat([input, torch.rand(1, 3, 32, 32)], dim=0)
names = list(n for n, _ in model.named_parameters())

# This is exactly the same code as in issue #49171
jac1 = jacobian(lambda *params: _stateless.functional_call(model, {n: p for n, p in zip(names, params)}, input), tuple(model.parameters()))
jac2 = jacobian(lambda *params: _stateless.functional_call(model, {n: p for n, p in zip(names, params)}, two_input), tuple(model.parameters()))
assert torch.allclose(jac1[0][0], jac2[0][0])
```

However, when I make the model slightly larger the assertion breaks down, which seem like it's due to rounding errors
```
class ResBasicBlock(nn.Module):
    def __init__(self, n_channels, n_inner_channels, kernel_size=3):
        super().__init__()

        self.conv1 = nn.Conv2d(n_channels, n_inner_channels, (kernel_size, kernel_size), padding=kernel_size // 2,
                               bias=False)
        self.conv2 = nn.Conv2d(n_inner_channels, n_channels, (kernel_size, kernel_size), padding=kernel_size // 2,
                               bias=False)
        self.norm1 = nn.BatchNorm2d(n_inner_channels)
        self.norm2 = nn.BatchNorm2d(n_channels)
        self.norm3 = nn.BatchNorm2d(n_channels)

    def forward(self, z, x=None):
        if x == None:
            x = torch.zeros_like(z)
        y = self.norm1(F.relu(self.conv1(z)))
        return self.norm3(F.relu(z + self.norm2(x + self.conv2(y))))

model = ResBasicBlock(3, 1)
input = torch.rand(1, 3, 32, 32)
two_input = torch.cat([input, torch.rand(1, 3, 32, 32)], dim=0)
names = list(n for n, _ in model.named_parameters())

# This is exactly the same code as in issue #49171
jac1 = jacobian(lambda *params: _stateless.functional_call(model, {n: p for n, p in zip(names, params)}, input), tuple(model.parameters()))
jac2 = jacobian(lambda *params: _stateless.functional_call(model, {n: p for n, p in zip(names, params)}, two_input), tuple(model.parameters()))
assert torch.allclose(jac1[0][0], jac2[0][0])

### Versions

```
Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.3 (x86_64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2)
CMake version: version 3.17.1
Libc version: N/A

Python version: 3.8.12 (default, Oct 12 2021, 06:23:56)  [Clang 10.0.0 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] functorch==0.1.0
[pip3] numpy==1.21.2
[pip3] torch==1.11.0
[pip3] torchaudio==0.11.0
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl    defaults
[conda] ffmpeg                    4.3                  h0a44026_0    pytorch
[conda] functorch                 0.1.0                    pypi_0    pypi
[conda] mkl                       2021.4.0           hecd8cb5_637    defaults
[conda] mkl-service               2.4.0            py38h9ed2024_0    defaults
[conda] mkl_fft                   1.3.1            py38h4ab4a9b_0    defaults
[conda] mkl_random                1.2.2            py38hb2f4e1b_0    defaults
[conda] numpy                     1.21.2           py38h4b4dc7a_0    defaults
[conda] numpy-base                1.21.2           py38he0bd621_0    defaults
[conda] pytorch                   1.11.0                  py3.8_0    pytorch
[conda] torchaudio                0.11.0                 py38_cpu    pytorch
[conda] torchvision               0.12.0                 py38_cpu    pytorch
```

**Comment by ezyang** [comment]
I'm guessing you don't want to work in double precision?

**Comment by QiyaoWei** [comment]
hmmm maybe I'm not very familiar with training models in double precision. Could you specify what I could do to test double precision? My (likely wrong) impression is that double precision only works on certain GPUs

**Comment by ezyang** [comment]
It's gonna be slow. But you can use this to check if there are deeper problems with the approach first. Do model.to(torch.double) to put the parameters in double

**Comment by QiyaoWei** [comment]
Oh ok that's what I thought. I still observe the same error though after changing these lines
```
model = ResBasicBlock(3, 1).double()
input = torch.rand(1, 3, 32, 32).double()
two_input = torch.cat([input, torch.rand(1, 3, 32, 32)], dim=0).double()
```

**Comment by zou3519** [comment]
The change in behavior is probably due to BatchNorm. BatchNorm normalizes across the batch elements, so calling BatchNorm on a single input A (with batch size 1) is different from calling BatchNorm on two inputs (A, B) and looking at the first result.

**Comment by QiyaoWei** [comment]
Right, that makes sense---would a good takeaway be to simply make the batch size 1 if I need to do the jacobian calculation? It feels unsettling that there's not a good way for me to verify whether the batch jacobian calculation is going correctly...

**Comment by ezyang** [comment]
Richard has nailed it. Your alternative is to pick a model architecture that isn't dependent on batch size. Hard to say without knowing what you're actually trying to do. In any case this isn't a PyTorch bug, so I'm gonna close this.
